File: Algorithm/Programmers/Python/Level 2/게임 맵 최단거리.py
# BFS는 최단거리를 구하는데 쓰인다. BFS로 처음 간 곳(1인 곳)은 무조건 최단거리이다.

from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    upDown = [0, 0, 1, -1]
    leftRight = [1, -1, 0, 0]
    
    que = deque([(0,0)])
    while que:
        x, y = que.popleft()
        
        for i in range(4):
            nx = x + leftRight[i]
            ny = y + upDown[i]
            
            if nx >= 0 and nx < m and ny >= 0 and ny < n and maps[ny][nx] == 1:
                que.append((nx, ny))
                maps[ny][nx] = maps[y][x] + 1

    if maps[n-1][m-1] == 1: # 도착점에 도달하지 못한 경우
        return -1
    else:
        return maps[n-1][m-1]

File: Algorithm/Programmers/Java/Level 2/게임_맵_최단거리.java
import java.util.*;

class 게임_맵_최단거리 {
    public int solution(int[][] maps) {
        int[] moveY = {0, 0, 1, -1};
        int[] moveX = {1, -1, 0, 0};
        
        int n = maps.length;
        int m = maps[0].length;
        int[][] visited = new int[n][m];
        visited[0][0] = 1;
        
        Deque<Integer[]> deque = new LinkedList();
        Integer[] coord = {0, 0};
        deque.offer(coord);
        
        while(!deque.isEmpty()) {
            Integer[] coor = deque.pollFirst();
            int y = coor[0];
            int x = coor[1];
            for(int i = 0; i < 4; i++) {
                int nextY = y + moveY[i];
                int nextX = x + moveX[i];
                if(nextY >= 0 && nextY < n && nextX >= 0 && nextX < m && maps[nextY][nextX] == 1
                    && visited[nextY][nextX] == 0) {
                    
                    visited[nextY][nextX] = visited[y][x] + 1;
                    Integer[] next = {nextY, nextX};
                    deque.offer(next);
                }
            }
        }
        
        if(visited[n - 1][m - 1] == 0) {
                return -1;
            }
            
        return visited[n - 1][m - 1];
    }
}